diff.py

```diff
@@ -2050,9 +2096,18 @@ def normalize_stack(row: str, arch: ArchSettings) -> str:
     return re.sub(arch.re_sprel, "addr(sp)", row)
 
 
-def imm_matches_everything(row: str, arch: ArchSettings) -> bool:
-    # (this should probably be arch-specific)
-    return "(." in row
+def imm_matches_everything(field: str, arch: ArchSettings) -> bool:
```

Comment:
Naming this `field` and including logic here that assumes it's run on an entire field is misleading, because this function is called with an entire instruction row (well, almost; the branch target part is stripped) by the existing caller.

This also means the below logic won't work for highlighting, only for scoring. You should be able to search for the `@\d+@(l|h|ha|sda21)\b` regex instead

Speaking of the existing caller: that uses `re_imm` which lists ha, h, l as the possible relocs. Needs to include sda21 too

Reply:
ohh that is not how I have been using this function, this is probably an error I made a long while back.
This is how I have been using it:


```
        for nf, of in zip(newfields, oldfields):
            # If the new field is a match to any symbol case
            if imm_matches_everything(nf, config.arch):
                # And the old field had a relocation, then ignore this mismatch
                if old_line.has_symbol:
```
 
On a specific code field, and that is why I renamed it, and that is why I made the logic the way I did because it was being used per field.

I will look into resoling this making a separate function, or changing the other places it is used...
the only other place it is used in the coloring code that I haven't done much with there.. maybe it's related to some other color issues

Reply:
are there even "@h" in PPC?  I haven't seen any.

vs "@ha" which I do see.

Reply:
once this is all finished... I'll circle back to the permuter and make all these fixes/updates there too.

Reply:
>function is called with an entire instruction row (well, almost; the branch target part is stripped) by the existing caller.

The only place where this happens is in the row parts coloring section, which I am reading and testing now to understand how it uses it... but I think we will want separate functions for the two places that call this function, since they use it differently.  The other function,  I need to understand and haven't touched is normalize_imms

>normalize_imms(branchless1, arch) == normalize_imms(branchless2, arch)

Reply:
yea at first glance, seems the `normalize_imms` is quite broken on ppc, even without any of my changes, or the updated regex.

Reply:
So first normalize_imms,
I don't think this can be solved with just a single regex, and I think we will need to break this down and check field by field, and ignore the mnemonics... but you know regex better than me so I'll let you give it a shot. here are some matching examples lhs and rhs.

```
lis     r3,lbl_803CEA7C@ha --- lis     r3,@10@ha
addi    r3,r3,lbl_803CEA7C@l --- addi    r3,r3,@10@l
lis     r3,lbl_803CEA70@ha --- lis     r3,@11@ha
addi    r3,r3,lbl_803CEA70@l --- addi    r3,r3,@11@l
addi    r5,0,lbl_804D3E50@sda21 --- addi    r5,0,@12@sda21
lfs     f0,lbl_804D9A18@sda21(0) --- lfs     f0,@13@sda21(0)

```

Also... now I have it setup so normalize_imms and imm_matches_everything, are the exact same behavior as they were before this PR... and both of these only affect inline coloring only but not scoring....  This is now a separate problem that I will work on and send another PR.  (However it will be slightly improved on ppc, because of the sda21 reloc fixes and li --> addi conversions).  So no need to let this particular issue hold up this current PR, that does the sda21 reloc stuff, and makes some fixes to scoring for all archs.

For this next PR (where I will focus on the inline coloring) I feel like for normalize imms we'll want to change to check this field by field... and same for imm_matches_everything, we will probably want to make that check field by field, or just use the other function I made for scoring.

Reply:
> are there even "`@h`" in PPC? I haven't seen any.

Not sure if they appear in practice tbh. Extrapolating from MIPS it feels like they'd be used if some code decides to materialize symbol addresses using bitor instead of addition.

Reply:
`field` -> `row`